### PR TITLE
Add the Cauchy kernel

### PR DESCRIPTION
**Pull request by laivtom**
Added an implementation and tests for the Cauchy kernel described by Basak (2008) in the article _A Least Square Kernel Machine With Box Constraints_.

### CODE REVIEW

File: src/mlpack/core/kernels/cauchy_kernel.hpp [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+namespace mlpack {
+namespace kernel {
+
+/**
+ * The Cauchy kernel. Given two vector @f$ x @f$, @f$ y @f$, and a bandwidth
+ * @f$ \sigma @f$ (set in the constructor),
+ *
+ * @f[
+ * K(x, y) = \frac{1}{1 + (\frac{|| x - y ||}{\sigma})^2}.
+ * @f]
```
Comment (rcurtin) [inline src/mlpack/core/kernels/cauchy_kernel.hpp]: Want to add a BibTeX reference here to the paper the kernel is from?

File: src/mlpack/core/kernels/cauchy_kernel.hpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+ */
+class CauchyKernel
+{
+ public:
+  /**
+   * Default constructor; sets bandwidth to 1.0.
+   */
+  CauchyKernel(): bandwidth(1.0)
```
Comment (rcurtin) [inline src/mlpack/core/kernels/cauchy_kernel.hpp]: Tiny picky style comment: can you change `CauchyKernel():` to `CauchyKernel() : `?

File: src/mlpack/core/kernels/cauchy_kernel.hpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+   * @tparam VecTypeB Type of second vector (arma::vec, arma::sp_vec).
+   * @param a First vector.
+   * @param b Second vector.
+   * @return K(a, b).
+   */
+  template<typename VecTypeA, typename VecTypeB>
+  double Evaluate(const VecTypeA& a, const VecTypeB& b)
+  {
+    return 1 / (1 + (
+        pow(metric::EuclideanDistance::Evaluate(a, b) / bandwidth, 2)));
```
Comment (rcurtin) [inline src/mlpack/core/kernels/cauchy_kernel.hpp]: I think this needs to be `std::pow()` to be properly portable.

**Comment by laivtom** [comment]
Thank you for the prompt review, @rcurtin :) Everything mentioned should be fixed now.

File: src/mlpack/core/kernels/cauchy_kernel.hpp [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+  /**
+   * Default constructor; sets bandwidth to 1.0.
+   */
+  CauchyKernel() : bandwidth(1.0)
+  { }
+
+  /**
+   * Construct the Cauchy kernel with a custom bandwidth.
+   */
+  CauchyKernel(double bandwidth) : bandwidth(bandwidth)
```
Comment (zoq) [inline src/mlpack/core/kernels/cauchy_kernel.hpp]: We could probably combine both constructors, and use ``bandwidth = 1.0`` as the default value.

**Pull request merged**